security/acme-client/src/opnsense/scripts/OPNsense/AcmeClient/upload_sftp.php

```diff
@@ -0,0 +1,1535 @@
+        }, $cmd));
+
+        $spec = [0 => ["pipe", "r"], 1 => ["pipe", "w"], 2 => ["pipe", "w"]];
+        $this->handle = proc_open($cmd, $spec, $pipes, $cwd, $env);
+
+        if (is_resource($this->handle)) {
+            $this->outputs = $pipes;
+            $this->inputs = [array_shift($this->outputs)];
+
+            foreach ($this->outputs as $stream)
+                stream_set_blocking($stream, false);
+
+            global $__open_processes;
+            $__open_processes[] = $this->handle;
+        } else {
+            logger()->error("Failed opening '$cmd' in '$cwd'");
+        }
+    }
+
+    public function __destruct()
+    {
+        $this->close();
+        if ($this->isRunning())
+            $this->close(true);
+    }
+
+    public function get($timeout = 5, $max_length = 8192, $ending = PHP_EOL)
+    {
+        $readables = array_filter($this->outputs, function ($stream) {
+            return is_resource($stream) && !feof($stream);
+        });
+
+        $micros = intval(($timeout - floor($timeout)) * 1000000);
+        $can_read = !empty($readables) && stream_select($readables, $w = [], $e = [], $timeout, $micros);
+        $stream = array_reduce(($can_read ? $readables : []), function ($a, $b) {
+            return is_resource($a) && !feof($a) ? $a : $b;
+        }, null);
+
+        return is_resource($stream)
+            ? stream_get_line($stream, $max_length, $ending)
+            : false;
+    }
+
+    public function put($data, $append = PHP_EOL)
+    {
+        if ($this->isRunning() && is_resource($stdin = $this->inputs[0]) && !feof($stdin)) {
+            fwrite($stdin, $data);
+            if ($append)
+                fwrite($stdin, $append);
+        }
+    }
+
+    public function closeInput()
+    {
+        if (!feof($stdin = $this->inputs[0])) fclose($stdin);
+    }
+
+    public function close($force = false)
+    {
+        global $__open_processes;
+
+        // Read up-to 10k remaining lines from STDOUT/ERR to release locks before closing.
+        for ($i = 0; ($line = $this->get(0)) && $i < 10000; $i++) {
+            logger()->error("WARN: process: $line");
+        };
+
+        if ($this->isRunning())
+            $this->exitCode = ($force ? proc_terminate($this->handle) : proc_close($this->handle));
+
+        if (!$this->isRunning() && in_array($this->handle, $__open_processes))
+            $__open_processes = array_diff($__open_processes, [$this->handle]);
+
+        return $this->exitCode;
+    }
+
+    public function isRunning()
+    {
+        $status = is_resource($this->handle) ? proc_get_status($this->handle) : false;
+        if (is_array($status)) {
+            if (!$this->exitCode && $this->exitCode !== 0 && !$status["running"])
+                $this->exitCode = $status["exitcode"];
+            return $status["running"];
+        }
+        return false;
+    }
+}
+
+
+// Calling main if we have been called via CLI
+if (isset($GLOBALS["argc"])) main();
```

Comment:
@fichtner @AdSchellevis  this may be used for Backup (Core) as well

Reply:
@jkellerer any reason for this uncommon call?

Reply:
No very good reason except that the script shouldn't just run when being included somewhere.

Reply:
Then it is common to have the library and the frontend (the code being called from CLI in different files. In such a case there is no "main" function and the content is just in the global scope of the file.

If multiple files contain a "main" function, you will get an exception that the function is already defined, so you will get a lot of trouble.

Reply:
That is true, I'd still prefer to organise the main code in the main function but I can remove the conditional for the call.